File: common_problems/extract/extract.ml
let rec extract n l =
  match n <= 0 with
  | true -> [[]]
  | false -> (
    match l with
    | [] -> []
    | h :: t ->
        let with_h =
          List.map
            (fun t' ->
              h :: t')
            (extract (n - 1) t)
        in
        let without_h =
          extract n t
        in
        with_h @ without_h )

let () = extract 3 ["a"; "b"; "c"; "d"]
    |> fun _ -> ()

File: common_problems/extract/extract.py
"""
tag: extract, distinct, binomial coefficient
"""

def extract(f, n, l):
    if n <=0:
        return [[]]
    if l == []:
        return []
    def with_h():
        h, *t = l
        return list(map(lambda lp: [h]+lp, extract("w/ ", n-1, t)))
    def without_h():
        return extract("w/o", n,l[1:])

    return with_h() + without_h()
print(extract("orig", 3,['a', 'b', 'c', 'd']))
